File: recipe/fraction.go
package recipe

import "fmt"

// Fraction is an integer divided by another integer.
type Fraction struct {
	Numerator, Denominator int
}

// IsWhole returns true if the simplified Fraction is a whole number.
func IsWhole(f Fraction) bool {
	return Simplify(f).Denominator == 1
}

// IsUndefined returns true if the denominator of the Fraction is 0.
func IsUndefined(f Fraction) bool {
	return f.Denominator == 0
}

// Simplify the Fraction by dividing the greatest common divisor of the
// numerator and denominator out.
func Simplify(f Fraction) Fraction {
	a, b := f.Numerator, f.Denominator
	for b != 0 {
		a, b = b, a%b
	}
	return Fraction{
		Numerator:   f.Numerator / a,
		Denominator: f.Denominator / a,
	}
}

func (f Fraction) String() string {
	f = Simplify(f)
	if f.Denominator == 1 {
		return fmt.Sprintf("%d", f.Numerator)
	}
	return fmt.Sprintf("%d/%d", f.Numerator, f.Denominator)
}

File: web/asset/fraction.js
// Fraction is an integer divided by another integer.
class Fraction {

  constructor(numerator, denominator) {
    this.numerator = numerator;
    this.denominator = denominator;
  }

  toString() {
    const f = simplify(this);
    if (f.denominator == 1) {
      return f.numerator.toString();
    }
    return f.numerator.toString() + '/' + f.denominator.toString();
  }

}

// isUndefined returns true if the denominator of the Fraction is 0.
function isUndefined(f) {
  return f.denominator == 0;
}

// isWhole returns true if the simplified Fraction is a whole number.
function isWhole(f) {
  return simplify(f).denominator == 1;
}

// simplify the Fraction by dividing the greatest common divisor of the
// numerator and denominator out.
function simplify(f) {
  let a = f.numerator;
  let b = f.denominator;
  while (b != 0) {
    const t = a;
    a = b;
    b = t % b;
  }
  return new Fraction(f.numerator/a, f.denominator/a);
}

// add the two Fractions and return the simplified form.
function add(a, b) {
  const numerator = b.denominator*a.numerator + a.denominator*b.numerator;
  const denominator = a.denominator * b.denominator;
  return simplify(new Fraction(numerator, denominator));
}
